packages/runtime/runtime-utils/src/serializer.ts

```diff
@@ -3,15 +3,99 @@
+//       yields an ~1.3x improvement (node 12 x64).
+function recursivelyReplace<TContext>(
+    context: TContext,
+    input: any,
+    replacer: (context: TContext, input: any) => any,
+) {
+    // If the current input is an IFluidHandle instance, replace this leaf in the object graph with
+    // the handle's serialized from.
+    const result = replacer(context, input);
+
+    // If the object has been replaced there is no need to descend (early exit).
+    // Note that this also early exits for NaN, which is fine since it's a leaf.
+    if (result !== input) {
+        return result;
+    }
+
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    let clone: object | undefined;
+
+    for (const key of Object.keys(input)) {
+        const value = input[key];
+        // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+        if (!!value && typeof value === "object") {
+            // Note: `input` has either been replaced (above) or must not contain circular references (as
+            //       object must be JSON serializable.)  Therefore, guarding against infinite recursion here
+            //       would only lead to a later error when attempting to stringify().
+            const replaced = recursivelyReplace(context, value, replacer);
+
+            // If the `replaced` object is different than the original `value` then the subgraph contained one
+            // or more replacements.  If this happens, we need to return a clone of the `input` object where
+            // the current property is replaced by the `replaced` value.
+            if (replaced !== value) {
+                // Lazily create a shallow clone of the `input` object if we haven't done so already.
+                clone = clone ?? (Array.isArray(input)
+                    ? input.slice(0)
+                    : { ...input });
+
+                // Overwrite the current property `key` in the clone with the `replaced` value.
+                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                clone![key] = replaced;
+            }
+        }
+    }
+
+    return clone ?? input;
+}
+
+/**
+ * Ensures the given 'handle' is bound to the provided 'bind' IFluidHandle and returns
+ * the handle's serialized form.
+ */
+function serializeHandle(handle: IFluidHandle, bind: IFluidHandle): ISerializedHandle {
+    bind.bind(handle);
+    return {
+        type: "__fluid_handle__",
+        url: handle.absolutePath,
+    };
+}
+
+/**
+ * If the given value is an IFluidHandle, returns the handle in serializable form.
+ * Otherwise returns the original object.
+ *
+ * Note that caller must ensure that 'value' is a non-null object.
+ */
+function encodeValue(bind: IFluidHandle, value: any) {
+    const handle = value.IFluidHandle;
+
+    return handle !== undefined
+        ? serializeHandle(handle, bind)
+        : value;
+}
```

Comment:
Defining 'encodeValue()' as a top-level function was required for the 1.3x perf gain mentioned above.